### ISSUE TITLE
Fix OSGi Manifest

### ISSUE THREAD

**Issue by hstaudacher** [ui4j/ui4j#28]
ui4j is not a valid OSGi bundle. Currently ui4j defines some OSGi Manifest headers. These are:

```
Bundle-Description: Web Automation for Java
Bundle-SymbolicName: com.ui4j.ui4j-all
Export-Package: com.ui4j.*
Bundle-Name: Ui4j - All
Bundle-Version: 1.2.0
```

The problematic part here is `com.ui4j.*`. Tools like Eclipse PDE have problems with the wildcard. I'm not even sure if this is valid in the OSGi specification (as I said, I'm not sure about this).

The packages should be exported separate and they should have also a version to be able to run different versions of ui4j in one OSGi container. Using the Maven OSGi bundle plugin this should not be a problem. 

Let me know if you need some help. Would be perfect if this could make it into the next release.

**Comment by hstaudacher** [comment]
Working on a path. Will open pull request in a few hours

**Comment by ui4j** [comment]
Hi, i am not familiar with OSGi. Pull request will be great.

**Comment by hstaudacher** [comment]
Harder as I thought. The heavy usage of the maven-shade plugin prevents an easy integration. The bundle plugin and the shade plugin are not playing well together. Not sure if I can create a solution today.

**Comment by ui4j** [comment]
For the version 2.0 it might be better to remove shade plugin for the **ui4j-webkit**. **ui4j-all** is already bundled with external dependencies (bytebuddy). If someone wants use fat jar **ui4j-all** could be used if not **ui4j-webkit** could be used with its external dependencies. Let me know if this change makes ui4j more OSGi friendly.

**Comment by hstaudacher** [comment]
Sorry I failed creating a proper OSGi integration. The api jar is not the problem. It only has one dependency to slf4j which is also an OSGi bundle. The webkit jar is more complicated. It has a lot of dependencies and not all are available as OSGi bundles e.g. xerces. So, removing the shading and using imports for the dependencies would result in a valid bundle but you couldn't easily resolve it's dependencies. I also tried to embed the depdencies which results in a huge fat jar which is not what is wanted ;).

So, one option is still left. Keeping the current shading and adding propper "export packages" manually. I think this solution is kind of hacky but easy to achieve. In the end it's up to you ;)

**Comment by ui4j** [comment]
maven-shade-plugin replaced with maven-bundle-plugin for **ui4j-api** and **ui4j-webkit** modules.
I try to create simple Eclipse RCP project but i cant fix ClassNotFoundException

[Controlsfx](https://bitbucket.org/controlsfx/controlsfx/issue/184/provide-osgi-meta-data-in-the-manifest) project has similar problem to run JavaFX on OSGi enviorments. As i understood its challenging task to run javafx on OSGi. Let me know if you could fix the ClassNotFound problems.

**Comment by hstaudacher** [comment]
Will test it tomorrow.

**Comment by hstaudacher** [comment]
I just builded and checked the manifests. As I said in the comment the imports are the problem. The webkit bundle has the imports;

```
Import-Package: com.sun.webkit.dom,com.sun.webkit.network,com.sun.webk
 it.network.about,com.sun.webkit.network.data,com.ui4j.api.browser;ver
 sion="[2.0,3)",com.ui4j.api.dialog;version="[2.0,3)",com.ui4j.api.dom
 ;version="[2.0,3)",com.ui4j.api.event;version="[2.0,3)",com.ui4j.api.
 interceptor;version="[2.0,3)",com.ui4j.api.util;version="[2.0,3)",com
 .ui4j.spi;version="[2.0,3)",javafx.application,javafx.beans,javafx.be
 ans.property,javafx.beans.value,javafx.concurrent,javafx.event,javafx
 .scene,javafx.scene.web,javafx.stage,javafx.util,net.bytebuddy,net.by
 tebuddy.dynamic,net.bytebuddy.dynamic.loading,net.bytebuddy.implement
 ation,net.bytebuddy.implementation.bind.annotation,net.bytebuddy.matc
 her,netscape.javascript,org.w3c.dom,org.w3c.dom.css,org.w3c.dom.event
 s,org.w3c.dom.html
```
One bad example is bytebuddy. It has also an invalid OSGi manifest which exports .* stuff. Also javafx come from the jvm, so no import is needed here. Not even sure if these imports can be satisfied.

**Comment by hstaudacher** [comment]
Also the all jar doe snot include bytebuddy anymore

**Comment by hstaudacher** [comment]
Hmm, not include is false. It does only not contain the implementation package from bytebuddy. So, it's unusable now.

**Comment by ui4j** [comment]
> One bad example is bytebuddy. It has also an invalid OSGi manifest which exports .* stuff.

There is a fixed [issue](https://github.com/raphw/byte-buddy/issues/5) about bytebuddy OSGi support.

>Hmm, not include is false. It does only not contain the implementation package from bytebuddy. So, it's unusable now.

So, Byte Buddy is not compatible with OSGi. Is it correct?

>Also javafx come from the jvm, so no import is needed here.

I am going to remove all jvm packages.

**Comment by hstaudacher** [comment]
Yes, if you take a look at the bytebuddy manifest you see this header: `Export-Package: !net.bytebuddy.jar.asm.*,net.bytebuddy.*`. OSGi doe snot know anything about `!` and so on. I assume these should have been the instruction for the Maven OSGi bundle plugin to not export the asm packages.

**Comment by raphw** [comment]
I updated Byte Buddy to support OSGi properly with version 0.6.8!

**Comment by hstaudacher** [comment]
nice! Thank you @raphw

**Comment by ui4j** [comment]
`BrowserFactory` depends on [ServiceLoader](https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html) feature but i am unable to use this feature for the OSGi environment. I try [this](http://blog.osgi.org/2013/02/javautilserviceloader-in-osgi.html) solution but it doesnt works for the Equinox.

Its possible to run Ui4j without ServiceLoader dependency, but an extra Java system parameter is required to fix ClassNotFoundException for the javafx classes.

Required parameter:
`-Dorg.osgi.framework.bundle.parent=app`

```java
public void start(BundleContext bundleContext) throws Exception {
    BrowserProvider provider = new WebKitBrowserProvider();
    BrowserEngine engine = provider.create();
    Page page = engine.navigate("https://www.google.com");
    page.show();
    Activator.context = bundleContext;
}
```
